README.md
# Leapfrog::CustomerScoring

[![Gem Version](https://badge.fury.io/rb/leapfrog-customer_scoring.png)](http://badge.fury.io/rb/leapfrog-customer_scoring)
[![Build Status](https://travis-ci.org/paulnsorensen/leapfrog-customer_scoring.png)](https://travis-ci.org/paulnsorensen/leapfrog-customer_scoring)
[![Dependency Status](https://gemnasium.com/paulnsorensen/leapfrog-customer_scoring.png)](https://gemnasium.com/paulnsorensen/leapfrog-customer_scoring)
[![Code Climate](https://codeclimate.com/github/paulnsorensen/leapfrog-customer_scoring.png)](https://codeclimate.com/github/paulnsorensen/leapfrog-customer_scoring)


Ruby API client for Leapfrog Online customer scoring

## Installation

Add this line to your application's Gemfile:

    gem 'leapfrog-customer_scoring'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install leapfrog-customer_scoring

## Usage

Require the gem.

    require 'leapfrog/customer_scoring'

Clients are instantiated with `http://internal.leapfrogonline/customer_scoring` as the default URL. 

    client = Leapfrog::CustomerScoring::Client.new

You may optionally pass a different URL to use.

    client = Leapfrog::CustomerScoring::Client.new("http://example.com")

To retrieve the scoring advice for a customer, you will call the `get_score(income, zipcode, age)` method. The response will be a `Hash` with two symbolized keys containing the `:propensity` and `:ranking` for that customer.

    advice = client.get_score("50000", "60621", "35")
    advice.inspect
    => "{:propensity=>0.26532, :ranking=>\"C\"}"

## Testing

To test this gem, you can checkout the master branch and run the `rake`
in your console.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
## v0.0.2

* Stylistic changes


## v0.0.1

* Initial release


spec/spec_helper.rb
require 'webmock/rspec'
require 'leapfrog/customer_scoring'

RSpec.configure do |config|
  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end

spec/leapfrog/customer_scoring/client_spec.rb
require 'spec_helper'

describe Leapfrog::CustomerScoring do
  describe "#initialize" do
    it "should have the correct default url" do
      client = Leapfrog::CustomerScoring::Client.new
      default_url = "http://internal.leapfrogonline.com/customer_scoring"
      expect(client.send(:url)).to eql(default_url)
    end

    it "should allow another url to be specified" do
      custom_url = "http://paulnsorensen.com"
      client = Leapfrog::CustomerScoring::Client.new(custom_url)
      expect(client.send(:url)).to eql(custom_url)
    end
  end

  describe "#get_score" do
    context "when successful" do
      before(:all) do
        test_url =  Leapfrog::CustomerScoring::Client::DEFAULT_URL
        test_url += "?age=35&income=50000&zipcode=60621"
        stub_request(:get, test_url).
        to_return(status: 200, body: '{"propensity": 0.26532,"ranking": "C"}')
        client = Leapfrog::CustomerScoring::Client.new
        @result = client.get_score("50000", "60621", "35")
      end

      it "should return a Hash" do
        expect(@result).to be_kind_of(Hash)
      end

      it "should only include 'propensity' and 'ranking" do
        expect(@result.keys.sort).to eql([:propensity, :ranking])
      end

      it "should have the correct value for 'propensity'" do
        expect(@result[:propensity]).to eql(0.26532)
      end

      it "should have the correct value for 'ranking'" do        
        expect(@result[:ranking]).to eql("C")
      end
    end

    context "when invalid url" do
      before(:all) do
        test_url =  Leapfrog::CustomerScoring::Client::DEFAULT_URL
        test_url += "?age=35&income=50000&zipcode=60621"
        stub_request(:get, test_url).
        to_return(status: [404, "Invalid URL"])
      end
      let(:client) { Leapfrog::CustomerScoring::Client.new }

      it "should raise an ResourceNotFound exception" do
        expect { client.get_score("50000", "60621", "35") }.
        to raise_error(Leapfrog::CustomerScoring::ResourceNotFound)
      end
    end

    context "when invalid input" do
      before(:all) do
        test_url =  Leapfrog::CustomerScoring::Client::DEFAULT_URL
        test_url += "?age=35&income=50000&zipcode=no"
        stub_request(:get, test_url).
        to_return(status: [422, "Invalid Input"])
      end
      let(:client) { Leapfrog::CustomerScoring::Client.new }

      it "should raise an InvalidInput exception" do
        expect { client.get_score("50000", "no", "35") }.
        to raise_error(Leapfrog::CustomerScoring::InvalidInput)
      end
    end

    context "when server error" do
      before(:all) do
        test_url =  Leapfrog::CustomerScoring::Client::DEFAULT_URL
        test_url += "?age=35&income=50000&zipcode=60621"
        stub_request(:get, test_url).
        to_return(status: [500, "Internal Server Error"])
      end
      let(:client) { Leapfrog::CustomerScoring::Client.new }

      it "should raise a ServerError exception" do
        expect { client.get_score("50000", "60621", "35") }.
        to raise_error(Leapfrog::CustomerScoring::ServerError)
      end
    end

    context "when timed out" do
      before(:all) do
        test_url =  Leapfrog::CustomerScoring::Client::DEFAULT_URL
        test_url += "?age=35&income=50000&zipcode=60621"
        stub_request(:get, test_url).
        to_timeout
      end
      let(:client) { Leapfrog::CustomerScoring::Client.new }

      it "should raise a ServerTimeout exception" do
        expect { client.get_score("50000", "60621", "35") }.
        to raise_error(Leapfrog::CustomerScoring::ServerTimeout)
      end
    end
  end
end

lib/leapfrog/customer_scoring/client.rb
module Leapfrog
  module CustomerScoring
    class Client
      DEFAULT_URL = "http://internal.leapfrogonline.com/customer_scoring"

      # You may initialize the Client with any url or just use the default
      # url of "http://internal.leapfrogonline.com/customer_scoring".
      # 
      #   client = Leapfrog::CustomerScoring::Client.new
      #
      # -- OR--
      #
      #   other_url = "http://example.com/customer_scoring"
      #   client = Leapfrog::CustomerScoring::Client.new(other_url)
      #
      # This is the url to which the client will make requests
      def initialize(url=DEFAULT_URL)
        @url = url
      end

      # Makes request to the initialized endpoint to retrieve customer
      # scoring advice. The <tt>income</tt>, <tt>zipcode</tt> and
      # <tt>age</tt> parameters are required. The return value is a
      # <tt>Hash</tt> with keys containing the <tt>:propensity</tt>
      # and <tt>:ranking </tt> for customer with the passed parameters
      #
      #   advice = client.get_score("50000", "60621", "35")
      #   advice.inspect
      #   => "{:propensity=>0.26532, :ranking=>\"C\"}"
      #
      def get_score(income, zipcode, age)
        params = {income: income, zipcode: zipcode, age: age}
        begin
          RestClient.get(url, params: params) do |response, request, result|
            case response.code
            when 200
              return ::JSON.parse(response, symbolize_names: true)
            when 422
              raise Leapfrog::CustomerScoring::InvalidInput
            when 404
              raise Leapfrog::CustomerScoring::ResourceNotFound
            when 500
              raise Leapfrog::CustomerScoring::ServerError
            else
              response.return!(request, result)
            end
          end
        rescue RestClient::RequestTimeout
          raise Leapfrog::CustomerScoring::ServerTimeout
        end
      end

      private
      attr_reader :url
    end
  end
end
